src/scoring/segmentation.py

```diff
@@ -0,0 +1,153 @@
+from pathlib import Path
+
+from loguru import logger as log
+from PIL import Image
+import torch
+from transformers import AutoImageProcessor, Mask2FormerForUniversalSegmentation
+
+try:
+    from typing import Optional, override
+except ImportError:
+    # for Python 3.9
+    from typing_extensions import Optional, override
+
+from src.scoring.scoring_method import ScoringMethod
+
+
+class Segmentation(ScoringMethod):
+    score_name = "segmentation_gvi"
+
+    def __init__(self):
+        """
+        Initializes processor and segmentation model
+        """
+        # Load the pretrained AutoImageProcessor from
+        # "facebook/mask2former-swin-large-cityscapes-semantic"
+        self.processor = AutoImageProcessor.from_pretrained(
+            "facebook/mask2former-swin-large-cityscapes-semantic"
+        )
+        # Set the device to GPU if available, otherwise use CPU
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        # Load the pretrained Mask2FormerForUniversalSegmentation
+        # "facebook/mask2former-swin-large-cityscapes-semantic"
+        model = Mask2FormerForUniversalSegmentation.from_pretrained(
+            "facebook/mask2former-swin-large-cityscapes-semantic"
+        )
+        # Move the self.model to the specified device (GPU or CPU)
+        self.model = model.to(device)
+
+    @override
+    def get_score(self, image_path: Path) -> Optional[float]:
+        """
+        Gets the segmentation GVI score for an image
+        Args:
+            image_path: Path to the image file
+
+        Returns: the index score as a float
+
+        """
+
+        try:
+            # Fetch and process the image
+            # image = Image.open(requests.get(image_path, stream=True).raw)
+            image = Image.open(image_path)
+            # Get the size of the image
+            width, height = image.size
+            # Crop the bottom 20% of the image to remove the band
+            # at the bottom of the panoramic image
+            bottom_crop = int(height * 0.2)
+            image = image.crop((0, 0, width, height - bottom_crop))
+            # Apply the semantic segmentation to the image
+            segmentation = self._segment_images(image)
+            # Cut panoramic image in 4 equal parts
+            # Crop the image and its segmentation based on
+            # the previously found road centers
+            images, pickles = self._crop_panoramic_images(image, segmentation)
```

Comment:
Why are the segmentations called "pickles"?

Reply:
https://docs.python.org/3/library/pickle.html